File: src/ch01/meetup/ruby/higher_order.rb
puts -> {
  compose = ->(f, g) {
    ->(x) {
      f.(g.(x))
    }
  }

  make_adder = ->(n) {
    ->(x) {
      x + n
    }
  }

  make_multiplier = ->(n) {
    ->(x) {
      x * n
    }
  }

  mul3add1 = compose.(make_multiplier.(3), make_adder.(1))
  mul3add1.(10)
}.()

File: src/ch01/meetup/cpp/higher_order.cc
#include <iostream>
using namespace std;

int main() {
	auto compose = [](auto f, auto g) -> auto {
		return [=](auto arg) auto {
			return f(g(arg));
		};
	};

	auto make_adder = [](auto n) -> auto {
		return [=](auto x) auto {
			return (x + n);
		};
	};

	auto make_multiplier = [](auto n) -> auto {
		return [=](auto x) auto {
			return (x * n);
		};
	};

	cout << compose(make_multiplier(3), make_adder(1))(10)
		 << '\n';
}
